actionpack/lib/action_dispatch/http/response.rb

```diff
@@ -410,7 +410,17 @@ def cookies
 
     def parse_content_type(content_type)
       if content_type
-        type, charset = content_type.split(/;\s*charset=/)
+        segments = content_type.split(/\s*;\s*/)
```

Comment:
How about extract more descriptive regex as `CONTENT_TYPE_PERSER`?

```ruby
    CONTENT_TYPE_PERSER = /\A(?<type>[^;\s]+)(?:\s*;\s*charset=(?<charset>[^;\s]+))?/ # :nodoc:

    def parse_content_type(content_type)
      if content_type && match = CONTENT_TYPE_PERSER.match(content_type)
        ContentTypeHeader.new(match[:type], match[:charset])
      else
        NullContentTypeHeader
      end
    end
```

Reply:
Looks pretty good. I'll change it soon.

By the way, https://tools.ietf.org/html/rfc7231#section-3.1.1.1 says `text/csv charset="utf-16"` is a valid representation of Content-Type. I don't know if we should support this quoted-string in this pull request. What do you think of this, @kamipo? If we support this type of quoted-string, I think we have no choice but to use Regexp here anyway.

Reply:
Yeah, I'm fine to support quoted-string, that is easer and no complex.

Reply:
hmm, I'm trying to support these patterns by Regexp parser, but it's not so simple...

```
text/csv; charset="utf-16"; header=present
text/csv; header=present; charset="utf-16"
```

Reply:
Does any client send the `text/csv; header=present; charset="utf-16"`?

Reply:
Anyway, it could make just one regex.

```ruby
    CONTENT_TYPE_PARSER = /\A(?<type>[^;\s]+)(?:.*;\s*charset=(?<quote>"?)(?<charset>[^;\s]+)\k<quote>)?/ # :nodoc:
```

Reply:
> Does any client send the text/csv; header=present; charset="utf-16"?

ah, a Rails app in our company has this kind of code... :innocent:
(To be accurate, our app returns `text/csv; header=present; charset=sjis`)

> Anyway, it could make just one regex.

Awesome! I pushed https://github.com/rails/rails/pull/35549/commits/dc3c2ec7873fcd7adc05824034c42a635cc2cee7, but slightly changed its Regexp like this:

```ruby
/\A(?<type>[^;\s]+)?(?:.*;\s*charset=(?<quote>"?)(?<charset>[^;\s]+)\k<quote>)?/
                   ^
              added this `?`
```

This is because its Content-Type might be half-cooked on `#parsed_content_type_header` and this make some existent tests to be failed. The code below is an example from test/dispatch/response_test.rb:

```rb
test "response charset and content type from railsish app" do
  @app = lambda { |env|
    ActionDispatch::Response.new.tap { |resp|
      resp.charset = "utf-16"
      # resp.headers['Content-Type'] #=> "; charset=utf-16" OMG ;(
      resp.content_type = Mime[:xml]
      resp.body = "Hello"
      resp.request = ActionDispatch::Request.empty
    }.to_a
  }

  get "/"
  assert_response :success

  assert_equal("utf-16", @response.charset)
  assert_equal(Mime[:xml], @response.content_type)

  assert_equal("application/xml; charset=utf-16", @response.headers["Content-Type"])
end
```

Reply:
> `# resp.headers['Content-Type'] #=> "; charset=utf-16" OMG ;(`

I see... nice catch :+1: